### PR TITLE
Ingester Processor

### PR DESCRIPTION
**Pull request by davit-y**
## Which problem is this PR solving?
- #929 

## Short description of the changes
- Unmarshal consumed messages into spans and write them to storage

### CODE REVIEW

File: cmd/ingester/app/processor/span_processor.go [hunk 1]
```diff
@@ -26,7 +31,30 @@ type SpanProcessor interface {
...
+
 // Message contains the fields of the kafka message that the span processor uses
 type Message interface {
 	Value() []byte
 }
+
+// NewSpanProcessor creates a new SpanProcessor
+func NewSpanProcessor(writer spanstore.Writer) SpanProcessor {
+	return &spanProcessor{
+		unmarshaller: kafka.NewProtobufUnmarshaller(),
```
Comment (vprithvi) [inline cmd/ingester/app/processor/span_processor.go]: I recommend making the unmarshaller configurable by using an `Option`.

That way, we can default to protobuf, but allow people to easily conveniently override it to anything else.

File: plugin/storage/kafka/unmarshaller.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+func (h *ProtobufUnmarshaller) Unmarshal(msg []byte) (*model.Span, error) {
+	newSpan := &model.Span{}
+	err := proto.Unmarshal(msg, newSpan)
+	return newSpan, err
+}
+
+// JSONUnmarshaller implements Unmarshaller
+type JSONUnmarshaller struct{}
+
+// NewJSONUnmarshaller constructs a ProtobufUnmarshaller
```
Comment (black-adder) [inline plugin/storage/kafka/unmarshaller.go]: if this was school, you'd be reprimanded

File: cmd/ingester/app/processor/span_processor.go [hunk 3]
```diff
@@ -26,7 +31,30 @@ type SpanProcessor interface {
...
+		unmarshaller: unmarshaller,
+		writer:       writer,
+	}
+}
+
+// Process unmarshals and writes a single kafka message
+func (s spanProcessor) Process(message Message) error {
+	mSpan, err := s.unmarshaller.Unmarshal(message.Value())
+	if err != nil {
+		return errors.Wrap(err, "cannot read message")
```
Comment (black-adder) [inline cmd/ingester/app/processor/span_processor.go]: this message is a bit too generic, maybe `could not unmarshal kafka message`?

File: cmd/ingester/app/processor/span_processor.go [hunk 4]
```diff
@@ -26,7 +31,30 @@ type SpanProcessor interface {
...
+	io.Closer
+}
+
 // Message contains the fields of the kafka message that the span processor uses
 type Message interface {
 	Value() []byte
 }
+
+// NewSpanProcessor creates a new SpanProcessor
+func NewSpanProcessor(writer spanstore.Writer, unmarshaller kafka.Unmarshaller) SpanProcessor {
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/ingester/app/processor/span_processor.go]: please don't use functional options. Pass an options struct if necessary.

[the remark this answers is not in the document]
Comment (davit-y) [reply cmd/ingester/app/processor/span_processor.go]: I will create a Params struct similar to the consumer

**Pull request merged**